### PR TITLE
NTuple Omit Prefix Implementation

### PR DESCRIPTION
**Pull request by ives1227**
This is the implementation of the recently merged spec https://github.com/OCFL/extensions/blob/main/docs/0007-n-tuple-omit-prefix-storage-layout.md.

### CODE REVIEW

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+public class NTupleOmitPrefixStorageLayoutConfig implements OcflExtensionConfig {
+
+	public static final String ZERO_PADDING_LEFT = "left";
+	public static final String ZERO_PADDING_RIGHT = "right";
+	public static final String ZERO_PADDING_NONE = "none";
+
+	private String delimiter;
+	private int tupleSize = 3;
+	private int numberOfTuples = 3;
+	private String zeroPadding = ZERO_PADDING_LEFT;
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java:28]: This would be better modeled as an enum.

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	/**
+	 * return true or false, indicates that the prefix-omitted, padded object
+	 * identifier should be reversed
+	 */
+	public boolean reverseObjectRoot() {
+		return reverseObjectRoot;
+	}
+
+	/**
+	 * The case-insensitive, delimiter marking the end of the OCFL object
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java:82]: Ack! I hate that I missed that this extension defines the delimiter as case-insensitive too. It makes the extension needlessly complex.

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	/**
+	 * the segment size (in characters) to split the digest into
+	 *
+	 * @param tupleSize
+	 *            - the segment size (in characters) to split the digest into
+	 * 
+	 */
+	public NTupleOmitPrefixStorageLayoutConfig setTupleSize(int tupleSize) {
+		if (tupleSize <= 0) {
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java:107]: The extension defines this as being between 1 and 32 inclusive. So, this could be rewritten as:

```java
this.tupleSize = Enforce.expressionTrue(tupleSize >=1 && tupleSize <= 32,
                tupleSize, "tupleSize must be between 1 and 32 inclusive");
```

**Review by pwinckles — CHANGES_REQUESTED** [review]
Thanks for submitting the extension!

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    @Override
+    public Class<? extends OcflExtensionConfig> getExtensionConfigClass() {
+        return NTupleOmitPrefixStorageLayoutConfig.class;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String mapObjectId(String objectId) {
```
[the remark this answers is not in the document]
Comment (pwinckles) [reply ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:109]: (Doing the character validation and split identification in a single pass is not required, and is unlikely to have much of a performance impact either way unless your ids are exceptionally long)

**Comment by ives1227** [comment]
@pwinckles I've addressed all of the above comments and recommitted.  Thanks so much for your timely review!

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        if (config == null) {
+            throw new OcflExtensionException("This extension must be initialized before it can be used.");
+        }
+        
+        if (!objectId.matches("\\A\\p{ASCII}*\\z")) {
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:111]: I like this, but please compile it to a static constant `Pattern` at the class level so that the regex does not need to be recompiled every time this method is called.

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        	throw new OcflExtensionException(String.format("This id %s must contain only ASCII characters.", objectId));
+        }
+        //Split by delimiter and get the last part
+        String id = objectId.toLowerCase();
+        int index = id.lastIndexOf(config.getDelimiter());
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:116]: The delimiter needs to be lowercased as well. See https://github.com/UW-Madison-Library/ocfl-java/blob/main/ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/FlatOmitPrefixLayoutExtension.java#L83

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+        if (index > -1) {
+            section = objectId.substring(index + config.getDelimiter().length());
+            baseObjectId = section;
+        }
+        else {
+        	throw new OcflExtensionException(String.format( "The delimiter %s cannot be found in %s.", config.getDelimiter(), objectId));
+        }
+        
+        if (OcflConstants.EXTENSIONS_DIR.equals(section) || section.isEmpty()) {
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:128]: You should probably remove the `section.isEmpty()` here as you are checking it in the next conditional. Up to you if you want to do the `length()` or `isEmpty()`.

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        
+        if (config.reverseObjectRoot()) {
+	        //Reverse the section
+        	section = new StringBuilder(section).reverse().toString();
+        }
+        //Add padding if needed and requested
+        if (section.length() < config.getTupleSize() * config.getNumberOfTuples()) {
+        	
+        	int paddingAmount = config.getTupleSize() * config.getNumberOfTuples();
+        	if (config.getZeroPadding().equals(NTupleOmitPrefixStorageLayoutConfig.ZeroPadding.LEFT)) {
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:144]: Enums can be compared by identity, so this is also valid:

```java
config.getZeroPadding() == NTupleOmitPrefixStorageLayoutConfig.ZeroPadding.LEFT
```

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        		section = "0".repeat(paddingAmount-section.length()) + section;
+        	}
+        	else if (config.getZeroPadding().equals(NTupleOmitPrefixStorageLayoutConfig.ZeroPadding.RIGHT)) {
+        		section = section + "0".repeat(paddingAmount-section.length());
+        	}
+        	//Throw runtime exception since we can't pad and there won't be enough characters for the pattern
+        	else {
+        		throw new OcflExtensionException(String.format("Zero padding is set to 'none' but %s is too short to follow the requested tuple pattern: %s.", section, config.toString()));
+        	}
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:153]: This case cannot exist because the enum must be left or right

File: pom.xml [hunk 10]
```diff
@@ -340,7 +339,7 @@
                 <artifactId>hamcrest-library</artifactId>
                 <version>2.2</version>
             </dependency>
-            <dependency>
+           	<dependency>
```
Comment (pwinckles) [inline pom.xml:342]: Can you revert all changes to this file?

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ * @author vcrema
+ * @since 2021-10-25
+ */
+public class NTupleOmitPrefixStorageLayoutConfig implements OcflExtensionConfig {
+
+	public enum ZeroPadding {
+		  LEFT,
+		  RIGHT
+	}
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java:46]: I don't think this will serialize to `left` and `right` as is required by the extension.

I _think_ it will work if you do the following, but you should confirm:

```java
public enum ZeroPadding {
    @JsonProperty("left")
    LEFT,
    @JsonProperty("right")
    RIGHT
}
```

File: ocfl-java-core/src/test/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageExtensionTest.java [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+import edu.wisc.library.ocfl.api.exception.OcflExtensionException;
+import edu.wisc.library.ocfl.api.exception.OcflInputException;
+import edu.wisc.library.ocfl.core.extension.storage.layout.NTupleOmitPrefixStorageLayoutExtension;
+import edu.wisc.library.ocfl.core.extension.storage.layout.config.NTupleOmitPrefixStorageLayoutConfig;
+
+/**
+ * @author vcrema
+ * @since 2021-10-25
+ */
+public class NTupleOmitPrefixStorageExtensionTest {
```
Comment (pwinckles) [inline ocfl-java-core/src/test/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageExtensionTest.java:41]: Can you add a test or two that exercises the case insensitive logic?

**Review by pwinckles — CHANGES_REQUESTED** [review]

**Comment by ives1227** [comment]
@pwinckles I'll get to these later tonight or tomorrow. Thanks!

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java (hunk 11 above)
Comment (ives1227) [reply ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/config/NTupleOmitPrefixStorageLayoutConfig.java:46]: I could not get the JsonProperty annotation to work properly but I was able to get this working in a different way.

File: ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        }
+        
+        if (!asciiPattern.matcher(objectId).matches()) {
+        	throw new OcflExtensionException(String.format("This id %s must contain only ASCII characters.", objectId));
+        }
+        //Split by delimiter and get the last part
+        String id = objectId.toLowerCase();
+        int index = id.lastIndexOf(config.getDelimiter().toLowerCase());
+        String section = objectId;
+        String baseObjectId = "";
```
Comment (pwinckles) [inline ocfl-java-core/src/main/java/edu/wisc/library/ocfl/core/extension/storage/layout/NTupleOmitPrefixStorageLayoutExtension.java:120]: These do not need to be initialized because their value is either assigned on lines 123-124 or an exception is thrown.

File: ocfl-java-itest/pom.xml [hunk 14]
```diff
@@ -31,12 +31,12 @@
     <parent>
         <groupId>edu.wisc.library.ocfl</groupId>
         <artifactId>ocfl-java-parent</artifactId>
-        <version>1.4.2-SNAPSHOT</version>
+        <version>1.4.1</version>
```
Comment (pwinckles) [inline ocfl-java-itest/pom.xml:34]: Please revert the changes to this file

File: ocfl-java-test/pom.xml [hunk 15]
```diff
@@ -31,12 +31,12 @@
     <parent>
         <groupId>edu.wisc.library.ocfl</groupId>
         <artifactId>ocfl-java-parent</artifactId>
-        <version>1.4.2-SNAPSHOT</version>
+        <version>1.4.1</version>
```
Comment (pwinckles) [inline ocfl-java-test/pom.xml:34]: Please revert the changes to this file

File: pom.xml [hunk 16]
```diff
@@ -31,7 +31,7 @@
 
     <groupId>edu.wisc.library.ocfl</groupId>
     <artifactId>ocfl-java-parent</artifactId>
-    <version>1.4.2-SNAPSHOT</version>
+    <version>1.4.1</version>
```
Comment (pwinckles) [inline pom.xml:34]: Please revert the changes to this file

**Review by pwinckles — CHANGES_REQUESTED** [review]
I pulled the PR and opened it in my IDE this time (the source of a few the nitpicks here) and noticed that you're using tabs whereas this project uses spaces. At some point I should get a linter setup, but, for now, please replace all tabs with four spaces.

Additionally, I tried to use the new extension and could not because the extensions spec file is missing. You need to add a copy of the extension spec to this directory https://github.com/UW-Madison-Library/ocfl-java/tree/main/ocfl-java-core/src/main/resources/ocfl-specs

**Comment by pwinckles** [comment]
I just committed some checkstyle rules that you can pickup if you rebase. I'm not a dogmatic code style person, but it is nice if fundamental things like spacing is consistent.

**Comment by ives1227** [comment]
@pwinckles I made the requested changes again.  Thanks for the checkstyles - that helped me clear up the discrepancies.

**Comment by pwinckles** [comment]
It appears that your PR is now including commits that were made to `main`. You need to rebase your PR so that it only includes your commits.

**Comment by ives1227** [comment]
@pwinckles I was just typing this to you.  I rebased on main because of the checkstyles and didn't realize it included all of these changes.  I'll fix this when I have a chance.

**Comment by pwinckles** [comment]
@ives1227 I'm hoping to get a release out this week or next. Do you think you'll have time to rebase this PR?

**Comment by ives1227** [comment]
@pwinckles I will try to get to this in the next day or two.  Between the holidays and work demands, I didn't have a chance to look at this.

**Comment by pwinckles** [comment]
You seem to have reverted the PR back to its original state, undoing all of the revisions that were made as part of the review process

**Comment by pwinckles** [comment]
I created an [example branch](https://github.com/UW-Madison-Library/ocfl-java/tree/ntuple-omit) of what this should look like. Note, that the tabs are still not fixed in that branch.

I created that checking out your branch and doing the following:

```shell
# Rebase to remove the three most recent commits
git rebase -i HEAD~4

# Add the upstream if you don't have it
git remote add upstream git@github.com:UW-Madison-Library/ocfl-java.git

# Rebase from upstream
git rebase upstream/main
```

That should produce the same branch as mine. Then, when you push to `origin` to update this PR you'll need to use the `--force` flag.

**Review by pwinckles — COMMENTED** [review]
Getting there... There are still pom files that are modified in this PR that should not be

**Comment by ives1227** [comment]
@pwinckles the files I had previously were 1.4.2-SNAPSHOT but when I rebased from your branch, they were 1.4.3-SNAPSHOT.  So, I should be using 1.4.3-SNAPSHOT instead?

**Comment by pwinckles** [comment]
> @pwinckles the files I had previously were 1.4.2-SNAPSHOT but when I rebased from your branch, they were 1.4.3-SNAPSHOT. So, I should be using 1.4.3-SNAPSHOT instead?

Yes, `1.4.3-SNAPSHOT` is correct

**Comment by ives1227** [comment]
@pwinckles I put the snapshot version back in.

**Review by pwinckles — APPROVED** [review]
Thank you for the submission!

**Comment by ives1227** [comment]
@pwinckles thank you!
